cleanlab/rank.py

```diff
@@ -152,14 +164,320 @@ def get_normalized_margin_for_each_label(
+    https://arxiv.org/abs/1911.00068
+
+    The purpose of this adjustment is to handle class imbalance.
+
+    Parameters
+    ----------
+    labels : np.array
+      A discrete vector of noisy labels, i.e. some labels may be erroneous.
+      *Format requirements*: for dataset with K classes, labels must be in {0,1,...,K-1}.
+
+    pred_probs : np.array (shape (N, K))
+      P(label=k|x) is a matrix with K model-predicted probabilities.
+      Each row of this matrix corresponds to an example x and contains the model-predicted
+      probabilities that x belongs to each possible class.
+      The columns must be ordered such that these probabilities correspond to class 0,1,2,...
+      `pred_probs` should have been computed using 3 (or higher) fold cross-validation.
+
+    Returns
+    -------
+    pred_probs_adj : np.array (float)
+    """
+
+    # Get expected (average) self-confidence for each class
+    confident_thresholds = get_confident_thresholds(labels, pred_probs)
+
+    # Subtract the class confident thresholds
+    pred_probs_adj = pred_probs - confident_thresholds
+
+    # Renormalize by shifting data to take care of negative values from the subtraction
+    pred_probs_adj += 1
+    pred_probs_adj /= pred_probs_adj.sum(axis=1)[:, None]
+
+    return pred_probs_adj
+
+
+def get_confident_thresholds(labels: np.array, pred_probs: np.array) -> np.array:
+    """Returns expected (average) "self-confidence" for each class.
+
+    The confident class threshold for a class j is the expected (average) "self-confidence" for class j.
+
+    See paper "Confident Learning: Estimating Uncertainty in Dataset Labels" by Northcutt et al.
+    https://arxiv.org/abs/1911.00068
+
+    Parameters
+    ----------
+    labels : np.array
+      A discrete vector of noisy labels, i.e. some labels may be erroneous.
+      *Format requirements*: for dataset with K classes, labels must be in {0,1,...,K-1}.
+
+    pred_probs : np.array (shape (N, K))
+      P(label=k|x) is a matrix with K model-predicted probabilities.
+      Each row of this matrix corresponds to an example x and contains the model-predicted
+      probabilities that x belongs to each possible class.
+      The columns must be ordered such that these probabilities correspond to class 0,1,2,...
+      `pred_probs` should have been computed using 3 (or higher) fold cross-validation.
+
+    Returns
+    -------
+    confident_thresholds : np.array (shape (K,))
+
+    """
+    confident_thresholds = np.array(
+        [np.mean(pred_probs[:, k][labels == k]) for k in range(pred_probs.shape[1])]
+    )
+    return confident_thresholds
+
+
+def score_label_quality(
+    labels: np.array,
+    pred_probs: np.array,
+    method: str = "self_confidence",
+    adj_pred_probs: bool = False,
+) -> np.array:
+    """Returns the label quality scores.
+
+    The scoring methods are heuristics to enable users to quickly rank order data to find potential label quality issues.
```

Comment:
would delete this sentence (especially 'heuristics' does not sound good). And replace with explanation above of what higher-vs-lower scores mean and intended range of scores.

Reply:
Done: 100783d9199110a66737f2d09e2fce762b500c45

Thanks